lib/spack/spack/test/spec_dag.py

```diff
@@ -69,6 +72,94 @@ def _mock(pkg_name, spec, deptypes=spack.alldeps):
     return _mock
 
 
+class MockPackage(object):
```

Comment:
I can't decide whether I like these or not.  They're obviously useful and the implementation here is pretty simple and clean.  But we also have `builtin_mock` -- should these be new packages there?  These have the advantage that the info you need to understand a test is entirely contained within the test, while `builtin.mock` is becoming pretty sprawling.  OTOH, `builtin.mock` actually tests package parsing and the Spack DSL.  So I dunno.

Probably it is ok to have both, at least for now.  So maybe `MockPackage` and friends should go in `conftest.py` so they're usable everywhere?

Reply:
> These have the advantage that the info you need to understand a test is entirely contained within the test, while builtin.mock

That's the main reason I did this and I think it's a really good one. I also use exactly the same new definitions when testing DAGs in https://github.com/LLNL/spack/pull/4595 and I think it will be useful for exercising any complex DAG structure.

> OTOH, builtin.mock actually tests package parsing and the Spack DSL. So I dunno.

I think that while this advocates for not eliminating builtin.mock, it does not mean this is redundant. Testing package parsing and Spack's DSL is important but doesn't have to be done in every test. In fact, arguably unit tests should be as targeted as possible and IMO more tests would benefit from using the `MockPackage` objects defined here instead of builtin.mock.